go/adbc/driver/databricks/external_browser_credentials.go

```diff
@@ -0,0 +1,307 @@
+	server := &http.Server{
+		Addr: fmt.Sprintf(":%d", port),
+		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			code := r.URL.Query().Get("code")
+			errorParam := r.URL.Query().Get("error")
+
+			if errorParam != "" {
+				errorDesc := r.URL.Query().Get("error_description")
+				fmt.Fprintf(w, browserErrorWithDescHtml, errorParam, errorDesc)
+				errorChan <- fmt.Errorf("oauth error: %s - %s", errorParam, errorDesc)
+				return
+			}
+
+			if code == "" {
+				fmt.Fprint(w, browserErrorGenericHtml)
+				errorChan <- fmt.Errorf("no authorization code received")
+				return
+			}
+
+			// Verify state parameter matches to prevent CSRF attacks
+			if r.URL.Query().Get("state") != state {
+				fmt.Fprintf(w, browserErrorWithDescHtml, "state mismatch", "state parameter did not match. Possible CSRF attack.")
+				errorChan <- fmt.Errorf("state mismatch - possible CSRF attack")
+				return
+			}
+
+			fmt.Fprint(w, browserSuccessHtml)
+			codeChan <- code
+		}),
+	}
+
+	// start local server
+	go func() {
+		if err := server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			errorChan <- fmt.Errorf("server error: %w", err)
+		}
+	}()
+
+	// wait for server to start up
+	time.Sleep(100 * time.Millisecond)
+
+	if err := openBrowser(ctx, authURL); err != nil {
+		return nil, err
+	}
+
+	// Wait for authorization code or 10 min timeout
+	var authCode string
+	select {
+	case authCode = <-codeChan:
+	case err := <-errorChan:
+		server.Shutdown(ctx)
+		return nil, fmt.Errorf("oauth callback error: %w", err)
+	case <-time.After(10 * time.Minute):
+		server.Shutdown(ctx)
+		return nil, fmt.Errorf("authentication timeout after 10 minutes - please ensure you complete the browser authentication flow")
+	case <-ctx.Done():
+		server.Shutdown(ctx)
+		return nil, fmt.Errorf("authentication cancelled: %w", ctx.Err())
+	}
+
+	// Shutdown server
+	server.Shutdown(ctx)
+
+	// Exchange code for tokens
+	return c.exchangeCodeForTokens(ctx, authCode, verifier, redirectURI)
+}
+
+// findAvailablePort finds an available port starting from the default
+func (c *ExternalBrowserCredentials) findAvailablePort() int {
+	// Try the default port first
+	if c.isPortAvailable(defaultPort) {
+		return defaultPort
+	}
+
+	// Try ports in a reasonable range
+	for port := defaultPort + 1; port < defaultPort+100; port++ {
+		if c.isPortAvailable(port) {
+			fmt.Printf("Port %d was in use, using port %d instead.\n", defaultPort, port)
+			return port
+		}
+	}
+
+	// If we can't find any available port, return the default and let it fail with a clear error
+	fmt.Printf("Warning: Could not find an available port, using default port %d.\n", defaultPort)
+	return defaultPort
+}
+
+// isPortAvailable checks if a port is available
+func (c *ExternalBrowserCredentials) isPortAvailable(port int) bool {
+	addr := fmt.Sprintf(":%d", port)
+	ln, err := net.Listen("tcp", addr)
+	if err != nil {
+		return false
+	}
+	ln.Close()
+	return true
```

Comment:
```suggestion
	if err == nil {
		ln.Close()
		return true
	} else {
	    return true
	}
```

Nit: save the boolean negation mental math when possible